hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java

```diff
@@ -327,6 +327,29 @@ private static AbsoluteTableIdentifier getAbsoluteTableIdentifier(Configuration
     return carbonSplits;
   }
 
+  private BitSet setMatchedPartitions(String partitionIds, CarbonTable carbonTable,
+      Expression filter) {
+    BitSet matchedPartitions = null;
+    if (null != partitionIds) {
+      String[] partList = partitionIds.replace("[","").replace("]","").split(",");
+      matchedPartitions = new BitSet(Integer.parseInt(partList[0]));
+      for (String partitionId : partList) {
+        matchedPartitions.set(Integer.parseInt(partitionId));
+      }
+    } else {
+      if (null != filter) {
+        PartitionInfo partitionInfo =
+            carbonTable.getPartitionInfo(carbonTable.getFactTableName());
+        if (null != partitionInfo) {
+          Partitioner partitioner = PartitionUtil.getPartitioner(partitionInfo);
+          matchedPartitions = new FilterExpressionProcessor()
+              .getFilteredPartitions(filter, partitionInfo, partitioner);
```

Comment:
I think we need modify getFilteredPartitions and FilterExpression to use partition id instead of array index.